<launch>
    <!-- Arguments that can be passed via command line-->
    <arg name="max_range" default="30.0" />
    <arg name="min_range" default="0.1" />
    <arg name="max_angular_correction_deg" default="20.0" />
    
    <!-- ICP parameters -->
    <arg name="max_correspondence_distance" default="0.5" />
    <arg name="max_iterations" default="30" />
    <arg name="transformation_epsilon" default="1e-8" />
    <arg name="fitness_epsilon" default="1e-6" />
    <arg name="voxel_grid_size" default="0.05" />

    <!-- Input scan topic -->
    <arg name="scan_topic" default="/scan" />
    
    <!-- TF frames -->
    <arg name="odom_frame" default="odom" />
    <arg name="base_frame" default="laser" />

    <!-- Launch the Simple LO node -->
    <node pkg="simple_lo" type="simple_lo_node" name="simple_lo_node" output="screen">
        <!-- Basic parameters -->
        <param name="max_range" value="$(arg max_range)" />
        <param name="min_range" value="$(arg min_range)" />
        <param name="max_angular_correction_deg" value="$(arg max_angular_correction_deg)" />
        
        <!-- ICP parameters -->
        <param name="max_correspondence_distance" value="$(arg max_correspondence_distance)" />
        <param name="max_iterations" value="$(arg max_iterations)" />
        <param name="transformation_epsilon" value="$(arg transformation_epsilon)" />
        <param name="fitness_epsilon" value="$(arg fitness_epsilon)" />
        <param name="voxel_grid_size" value="$(arg voxel_grid_size)" />
        
        <!-- Remap topics -->
        <remap from="/scan" to="$(arg scan_topic)" />
        
        <!-- TF frames -->
        <param name="odom_frame" value="$(arg odom_frame)" />
        <param name="base_frame" value="$(arg base_frame)" />
    </node>

    <!-- Optional: Launch RViz for visualization -->
    <arg name="rviz" default="true" />
    <group if="$(arg rviz)">
        <node pkg="rviz" type="rviz" name="rviz" 
              args="-d $(find simple_lo)/rviz/simple_lo.rviz" />
    </group>

</launch> 
